src/jace/util/util.py

```diff
@@ -0,0 +1,30 @@
+# JaCe - JAX Just-In-Time compilation using DaCe (Data Centric Parallel Programming)
+#
+# Copyright (c) 2024, ETH Zurich
+# All rights reserved.
+#
+# SPDX-License-Identifier: BSD-3-Clause
+
+from __future__ import annotations
+
+from collections.abc import Iterable
+from typing import Any
+
+
+def ensure_iterability(
+    x: Any,
+    ign_str: bool = True,
+) -> Iterable[Any]:
+    """Ensures that `x` is iterable.
+
+    By default strings are _not_ considered iterable.
+
+    Args:
+        x:          To test.
+        ign_str:    Ignore that a string is iterabile.
+    """
+    if ign_str and isinstance(x, str):
+        x = [x]
+    elif isinstance(x, Iterable):
+        pass
+    return x
```

Comment:
Ok, there are two aspects here: one is checking if an object is an iterable instance which is not a string and the second is to wrap an element into a single-element sequence. I was talking only about the first one, in which case the `ign_str` parameter doesn't make sense because it has to be `True` for this function to be useful, since when it is `False` it'd be equivalent to `isinstance(x, Iterable)`. 

So what I proposed was to define the function as a typeguard (https://typing.readthedocs.io/en/latest/spec/narrowing.html#typeguard) for a new `Iterable` subtype (e.g. `NonStringIterable`) representing Iterable instances that are not strings. Something like:

```python
class NonStringIterable(Iterable): ...

def is_non_string_iterable(val: Iterable) -> Typeguard[NonStringIterable]:
    return isinstance(val, Iterable) and not isinstance(val, str)
```
A slightly more _magic_ alternative would be to directly implement a subclashook in the custom Iterable subclass:
```python
class NonStringIterable(Iterable):
    @classmethod
    def __subclasshook__(cls, C):
        return not isinstance(C, str) and super().__subclasshook__(C)
```
For the second aspect, wrapping atomic elements into single-element sequences, you can define a similar function as the current one but again I'd drop the `ign_str` argument since it's currently unused (and most likely it will always be that way), and I'd add the right overloads to accurate express the types, as for example: 

```python
@overload
def as_sequence(value: str) -> Iterable[str]: ...

@overload
def as_sequence(value: Iterable[T]) -> Iterable[T]: ...

@overload
def as_sequence(value: T) -> Iterable[T]: ...

def as_sequence(value: T | Iterable[T]) -> Iterable[T]:
    return [value] if not isinstance(value, Iterable) or isinstance(value, str) else value
```

Note: all the code here is basically untested pseudocode

Reply:
Thanks I was not aware of this.
I made an implementation, of your code and it worked, kind of, but you should take a look as it is very likely not fully correct.